ui/ui-components/client/client.ts

```diff
@@ -1,104 +1,85 @@
-import type { ErrorHandler } from "../eventHandlers/errorHandler";
-import { isBrowser } from "../utils/isBrowser";
+import { errorHandler } from "../eventHandlers";
 import { getRedirectFromErrorCode } from "../utils/getRedirectFromErrorCode";
+import { isBrowser } from "../utils/isBrowser";
 
-interface ClientCacheObject<T = unknown> {
-  locked: boolean;
-  data: T;
-  createdAt: number;
-}
+import { ClientCache, ClientCacheGetObject } from "./clientCache";
 
-interface ClientCacheGetObject<T = unknown> {
-  cacheData: T;
-  unlock: (data?: T) => void;
+interface ClientRequestOptions {
+  useCache: boolean;
 }
 
-export class ClientCache {
-  cache: { [key: string]: ClientCacheObject };
-  ttl: number;
+const API_VERSION = process.env.API_VERSION || "v1";
+const WEB_URL = process.env.WEB_URL ?? "";
+const SERVER_TOKEN = process.env.SERVER_TOKEN;
 
-  constructor(ttl = 5000) {
-    this.ttl = ttl;
-    this.cache = {};
-  }
+const getCsrfToken = () =>
+  globalThis?.localStorage
+    ? window.localStorage.getItem("session:csrfToken")
+    : undefined;
 
-  async sleep(wait = 100) {
-    return new Promise((resolve) => {
-      setTimeout(resolve, wait);
-    });
-  }
+const apiInstance = axios.create({
+  baseURL: `${WEB_URL}/api/${API_VERSION}`,
+  withCredentials: true,
+  headers: {
+    Accept: "application/json",
+    "Content-Type": "application/json",
+    "X-Grouparoo-Client": `${PackageJSON.name}-v${PackageJSON.version}`,
+  },
+});
 
-  async get<T = unknown>(key: string): Promise<ClientCacheGetObject<T>> {
-    if (!this.cache[key]) {
-      this.cache[key] = { locked: false, data: undefined, createdAt: 0 };
-    }
+apiInstance.interceptors.response.use(
+  (response) => {
+    if (response.data && response.data.error) {
+      const error = response.data.error.message
+        ? response.data.error.message
+        : response.data.error;
 
-    let waitCounter = 0;
-    while (this.cache[key]?.locked && waitCounter < 15) {
-      await this.sleep(300);
-      waitCounter++;
+      throw error;
     }
-
-    const now = new Date().getTime();
-    if (this.cache[key]?.data) {
-      if (this.cache[key].createdAt + this.ttl < now) {
-        delete this.cache[key];
-        this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
-      }
+    return response;
+  },
+  (error) => {
+    if (error.response && error.response.data && error.response.data.error) {
+      const err =
+        error.response?.data?.error?.message ??
+        error.response?.data?.error ??
+        error;
+
+      const newError = new Error(err);
+      newError["code"] = error.response?.data?.error?.code;
+      throw newError;
     } else {
-      this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
+      throw error;
     }
+  }
+);
 
-    const unlock = (data?: T) => {
-      this.cache[key] = {
-        data,
-        createdAt: new Date().getTime(),
-        locked: false,
-      };
-    };
+if (isBrowser()) {
```

Comment:
The biggest change is here: All the client-side logic is consolidated here instead of the `Client` class. I tried a few different approaches and this is the cleanest I could come up with. The other is to apply it to the `useApi` hook somehow but I found it too messy.